test/MarketInvariants.t.sol

```diff
@@ -0,0 +1,87 @@
+/* solhint-disable contract-name-camelcase, func-name-mixedcase, reason-string */
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.17;
+import "@/test/helpers/test.sol";
+import "@/test/helpers/market.sol";
+import "@/test/helpers/MarketHandler.sol";
+import "@/contracts/Certificate.sol";
+import "@/contracts/Removal.sol";
+import "forge-std/InvariantTest.sol";
+import "forge-std/console2.sol";
+import {console} from "forge-std/console.sol";
+
+contract MarketInvariantTest is InvariantTest, Global {
+  MarketHandler marketHandler;
+  Certificate internal _certificate;
+  Removal internal _removal;
+
+  function setUp() external {
+    marketHandler = new MarketHandler();
+    _certificate = Certificate(marketHandler.getCertificateAddress());
+    _removal = Removal(marketHandler.getRemovalAddress());
+    targetContract(address(marketHandler)); // only target the handler
+  }
+
+  function invariant_sumOfPurchaseAmounts() external {
```

Comment:
like test in forge tests that we're familiar with, invariant tests are indicated by prefixing the test name with invariant_

here I have just one invariant defined -- which asserts that the sum of purchase amounts in the market should always be equal to the sum of the certificate contracts balances of all removal ids that it owns any amount of, less its own reported nrtDeficit.

this invariant relies on a "ghost variable" in the handler contract, which keeps track of which removals the certificate has any ownership of, so that a sum can be performed while checking the invariant.